packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx

```diff
@@ -1,57 +1,68 @@
+    const { client, common } = useStore();
+    const { is_language_changing } = common;
+    const { has_enabled_two_fa, setTwoFAChangedStatus, setTwoFAStatus } = client;
     const [is_success, setSuccess] = React.useState(false);
-    const [is_ready_for_verification, setReadyForVerification] = React.useState(false);
-    const button_text = is_enabled ? localize('Disable') : localize('Enable');
-    const formik_ref = React.useRef();
-    let enable_response;
+
+    const button_text = has_enabled_two_fa ? localize('Disable') : localize('Enable');
+    const formik_ref = React.useRef<FormikProps<TDigitFormValues>>(null);
+    let enable_response: TResponse;
 
     const initial_form = {
         digit_code: '',
     };
 
     React.useEffect(() => {
         if (is_language_changing) {
-            formik_ref.current.setFieldTouched('digit_code');
+            formik_ref.current?.setFieldTouched('digit_code');
         }
     }, [is_language_changing]);
 
-    const validateFields = async values => {
+    const validateFields = async (values: TDigitFormValues) => {
         const digit_code = values.digit_code;
         if (!digit_code) {
             return { digit_code: localize('Digit code is required.') };
         } else if (!(digit_code.length === 6)) {
             return { digit_code: localize('Length of digit code must be 6 characters.') };
         } else if (!/^[0-9]{6}$/g.test(digit_code)) {
             return { digit_code: localize('Digit code must only contain numbers.') };
-        } else if (is_ready_for_verification) {
-            if (formik_ref.current.isValid) {
-                const totp_action = is_enabled ? 'disable' : 'enable';
-                enable_response = await WS.authorized.accountSecurity({
-                    account_security: 1,
-                    totp_action,
-                    otp: values.digit_code,
-                });
-                if (enable_response.error) {
-                    const { code, message } = enable_response.error;
-                    if (code === 'InvalidOTP')
-                        return { digit_code: localize("That's not the right code. Please try again.") };
-                    return { digit_code: message };
-                }
-            } else {
-                return { digit_code: localize("That's not the right code. Please try again.") };
+        } else if (formik_ref.current?.isValid) {
+            const totp_action = has_enabled_two_fa ? 'disable' : 'enable';
+            enable_response = await WS.authorized.accountSecurity({
+                account_security: 1,
+                totp_action,
+                otp: values.digit_code,
+            });
+            if (enable_response?.error) {
+                const { code, message } = enable_response.error;
+                if (code === 'InvalidOTP')
+                    return { digit_code: localize("That's not the right code. Please try again.") };
+                return { digit_code: message };
```

Comment:
In the previous code, we were displaying "That's not the right code. Please try again."  in the else condition for (formik_ref.current?.isValid) 
 where are we setting this in new code?

Reply:
We actually don't need it anymore. In the previous implementation we were checking if `is_ready_for_verification` is true and then only we make request call. In this implementation we remove the usage of that useState as its redundant. We can check if above conditions are fulfilled before sending OTP to BE by simply looking at formik error obj. So, basically If error happens we display `That's not the right code. Please try again.`. Also, there is no other case possible when we show this msg except for invalid OTP. Basically, previous implementation's else statement was redundant.

Reply:
@utkarsha-deriv can you check what will happen if (formik_ref.current?.isValid) is false in the new implementation?

Reply:
if (formik_ref.current?.isValid) is false it means either of the first three validation cases fails, so error msg is shown accordingly. One more case is possible, that is when first 3 validation cases pass and we enter last condition. So, user can send OTP but if BE responds back with error we populate the error obj and (formik_ref.current?.isValid) becomes false again. This prevents the user from submitting the same invalid OTP again.

Reply:
On testing the flow, I think its better to introduce a condition along with (formik_ref.current?.isValid) to make sure we make the call to BE only when the user clicks on the button. Currently, its making call on any click event not specific to enable/disable btn. Will fix this now. 🙏